Adding VCALM Threat Model (#655)
w3c/vcalm: threat_model/README.md

# Respec-Threats

A ReSpec rendering library for threat models

> 

Reviewer: We need to factor this out into a `respec-threat-model` library. Ivan, can you please set up a repo at W3C for us to put this library into?

Reviewer 2: I think you forgot to tag @iherman here.

Reviewer 3: (Indeed he did...:-)

@msporny @kezike, please give me the exact name of the repository you wish to have? To be proactive, does `vcalm-thread-model` work for you?

Reviewer 4: > We need to factor this out into a `respec-threat-model` library. Ivan, can you please set up a repo at W3C for us to put this library into?

There is a separate respect threats library. 
https://github.com/legReq/respec-threats

I'm not sure what you're proposing.  You want the W3C to take ownership of that repo?

Reviewer: Yes, my expectation was that all respec extensions would end up under W3C at some point. People got very cranky with me when I had other respec extensions in my personal Github repo. We could just migrate that `respec-threats` repo? 

I thought we could start moving that sooner than later. Were you not expecting to do that, @jandrieu ?

Author: Created an issue in the legreq/respec-threats repo to capture this discussion: https://github.com/LegReq/respec-threats/issues/6

# Use 

To use this library, add the library files to your repo, then add threat

> definition files, and finally and update your spec's html to use them.

Reviewer 5: This appears to be a copy-and-paste from somewhere, to which this update should be retrofitted.
```suggestion
definition files, and finally update your spec's HTML to use them.
```

## Library Files

The library files you'll want to add:

1. `threats/threat-model.js` the javascript that handles parsing the threats and
   rendering them in the spec.
2. `threats/threat-model.css` the css file that handles the layout for the

>    threats and table of contents. 

Reviewer 5: ```suggestion
1. `threats/threat-model.js` the JavaScript that handles parsing the threats and
   rendering them in the spec.
2. `threats/threat-model.css` the CSS file that handles the layout for the
   threats and table of contents. 
```

## Threat Definition Files

The threat definition files you'll want to add:

> 1. Individual threat.js files (in the threats folder in the example) as

Reviewer 5: ```suggestion
1. Individual `threat.js` files (in the `threats` folder in the example) as
```

   described in the example spec.
2. `threats/outline.js` This defines the categories for organizing both the
   table of contents and the threat details themselves. Any number of categories
   can be defined, with any number of threats in each, with any name for each
   category. The outline uses just the short form ID for each threat, e.g., "T1"
   or "T5".

## HTML updates

> 1. Add a script tag for threat-model.js. This MUST be before the Respec Config

Reviewer 5: ```suggestion
1. Add a `script` tag for `threat-model.js`. This MUST be before the Respec Config
```

   block *and* the other Respec-Threats javascript tags.

```html
  <script src="threats/threat-model.js"></script> <!-- must precede the respec config-->
```

1. Add suitable selectors for the Table of Contents and Threat Details sections

>    in the Respec Config (see example). AND add `window.ThreatModel.render` to

Reviewer 5: ```suggestion
   in the Respec Config (see example) _AND_ add `window.ThreatModel.render` to
```

   the preprocessing directive. Note that the Table of Contents section
   *MUST* have an id.

   ```JavaScript
      threatModelTocSelector: "#threat-model-toc",
      threatModelDetailsSelector: "#threat-model-details",
      preProcess: [window.ThreatModel.render]
   ```

> 1. Add a link tag to bring in the threat-model.css, e.g.,

Reviewer 5: ```suggestion
1. Add a `link` tag to bring in the `threat-model.css`, e.g.,
```

```html
  <link rel="stylesheet" href="threats/threat-model.css" />
```

> 1. Add script tag for the outline.js file

Reviewer 5: ```suggestion
1. Add `script` tag for the `outline.js` file
```

```javascript
<script src="threats/outline.js"></script> 
```

1. Finally, add script tags for each of the individual threat definition files.

>    The file names will vary for your spec, but here's the list in the example spec.

Reviewer 5: ```suggestion
1. Finally, add `script` tags for each of the individual threat definition files.
   The file names will vary in your specification, but here's the list in the example.
```

```html
  <script src="threats/t1-imposter-websites.js"></script>
  <script src="threats/t2-network-surveillance.js"></script>
  <script src="threats/t3-network-corruption.js"></script>
  <script src="threats/t4-fraudulent-certificate.js"></script>
  <script src="threats/t5-key-compromise.js"></script>
  <script src="threats/t6-library-compromise.js"></script>
  <script src="threats/t7-quantum-cryptography-attacks.js"></script>
  <script src="threats/t8-harvest-attacks.js"></script>
  <script src="threats/t9-flawed-cryptosuite.js"></script>
  <script src="threats/t10-the-internet-protocol-suite.js"></script>
```

Note: We recommend keeping the contribution directions somewhere in your
repository, as it will help the community contribute threats to your
specification. It need not be in the body of the spec itself, but it should be a

> good run down of what contributors need to do to help with threat modeling.

Reviewer 5: ```suggestion
specification. It need not be in the body of the specification itself, but it should
be a good run down of what contributors need to do to help with threat modeling.
```
